src/expression.c

```diff
@@ -118,7 +118,12 @@ Expression *getRightThis(Loc loc, Scope *sc, AggregateDeclaration *ad,
                     {
                         //printf("rewriting e1 to %s's this\n", f->toChars());
                         n++;
+#ifdef IN_GCC
+                        if (n > 1)
+                            e1 = new VarExp(loc, f->vthis);
+#else
                         e1 = new VarExp(loc, f->vthis);
+#endif
```

Comment:
It looks like ``AggregateDeclaration::makeNested`` assigns the Outer class as the parent ``vthis`` before all closure information is available.  Not that having all closure information available helps in any way.